### PR TITLE
Add enhancement to find

### LINKED ISSUE #72: Implement find command to handle finding by both name and classId

### PR DESCRIPTION
**Pull request by hooyushing**
fixes #72

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 1]
```diff
@@ -41,6 +49,17 @@ public FindCommand parse(String args) throws ParseException {
...
+
+        if (nameClassId.length < 2 || nameClassId[1].trim().isEmpty()) {
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+        }
+
+        return nameClassId[0].trim().split("\\s+");
     }
 
     private String[] parseName(String args) throws ParseException {
```
Comment (sumomomomomo) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:65]: Shouldn't parseName and parseClassId be abstracted out to ParserUtil? Referring to AddCommandParser

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 2]
```diff
@@ -28,6 +29,13 @@ public FindCommand parse(String args) throws ParseException {
...
+        if (trimmedArgs.contains("n/") && trimmedArgs.contains("c/")) {
+            String[] nameKeywords = parseNameClassId(trimmedArgs);
+            String[] classIdKeywords = parseClassId(trimmedArgs);
+            return new FindCommand(new NameAndClassIdContainsKeywordsPredicate(Arrays.asList(nameKeywords),
+                    Arrays.asList(classIdKeywords)));
+        }
+
         if (trimmedArgs.contains("n/")) {
             String[] nameKeywords = parseName(trimmedArgs);
             return new FindCommand(new NameContainsKeywordsPredicate(Arrays.asList(nameKeywords)));
```
Comment (sumomomomomo) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:41]: Shouldn't all of these be abstracted out of FindCommandParser? Referring to AddCommandParser

File: src/main/java/seedu/address/model/person/NameAndClassIdContainsKeywordsPredicate.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof NameAndClassIdContainsKeywordsPredicate)) {
+            return false;
+        }
+
+        NameAndClassIdContainsKeywordsPredicate otherClassIdContainsKeywordsPredicate =
```
Comment (sumomomomomo) [inline src/main/java/seedu/address/model/person/NameAndClassIdContainsKeywordsPredicate.java:50]: nit: `otherClassIdContainsKeywordsPredicate` should be `otherNameAndClassIdContainsKeywordsPredicate`

File: src/main/java/seedu/address/model/person/NameAndClassIdContainsKeywordsPredicate.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public NameAndClassIdContainsKeywordsPredicate(List<String> nameKeywords, List<String> classIdKeywords) {
+        this.nameKeywords = nameKeywords;
+        this.classIdKeywords = classIdKeywords;
+    }
+
+    @Override
+    public boolean test(Person person) {
+        return classIdKeywords.stream()
+                .anyMatch(keyword -> regexMatch(person.getClassId().value , keyword))
```
Comment (sumomomomomo) [inline src/main/java/seedu/address/model/person/NameAndClassIdContainsKeywordsPredicate.java:30]: nit: ` , keyword` should be `, keyword`

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 5]
```diff
@@ -80,6 +100,17 @@ public void execute_zeroKeywords_noPersonFound() {
         assertEquals(Collections.emptyList(), model.getFilteredPersonList());
     }
 
+    @Test
+    public void executeNoPersonsFoundClassId() {
```
Comment (sumomomomomo) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:104]: Naming convention

**Review by sumomomomomo — COMMENTED** [review]
Overall good but can you abstract out the things from FindCommandParser to follow convention (of all the other Parser classes)

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java (hunk 2 above)
Comment (hooyushing) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java:41]: I think this is fine for the if statements and the block

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java (hunk 1 above)
Comment (hooyushing) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java:65]: Fixed

File: src/main/java/seedu/address/model/person/NameAndClassIdContainsKeywordsPredicate.java (hunk 3 above)
Comment (hooyushing) [reply src/main/java/seedu/address/model/person/NameAndClassIdContainsKeywordsPredicate.java:50]: Fixed

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 6]
```diff
@@ -28,6 +29,13 @@ public FindCommand parse(String args) throws ParseException {
                     String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
         }
 
+        if (trimmedArgs.contains("n/") && trimmedArgs.contains("c/")) {
```
Comment (sumomomomomo) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:32]: Don't use magic literals here

File: src/test/java/seedu/address/model/person/NameAndClassIdContainsKeywordsPredicateTest.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        List<String> keywords = List.of("keyword1", "keyword2");
+        List<String> keywords2 = List.of("keyword3", "keyword4");
+        NameAndClassIdContainsKeywordsPredicate predicate =
+                new NameAndClassIdContainsKeywordsPredicate(keywords, keywords2);
+
+        String expected = NameAndClassIdContainsKeywordsPredicate.class.getCanonicalName()
+                + "{Name keywords=" + keywords + ", ClassId keywords=" + keywords2 + "}";
+        assertEquals(expected, predicate.toString());
+    }
+}
```
[the remark this answers is not in the document]
Comment (hooyushing) [reply src/test/java/seedu/address/model/person/NameAndClassIdContainsKeywordsPredicateTest.java:99]: Fixed all naming conventions for find tests

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java (hunk 6 above)
Comment (hooyushing) [reply src/main/java/seedu/address/logic/parser/FindCommandParser.java:32]: Fixed

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 8]
```diff
@@ -1,13 +1,15 @@
...
 import static java.util.Objects.requireNonNull;
+import static seedu.address.logic.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
 
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.Set;
 
 import seedu.address.commons.core.index.Index;
 import seedu.address.commons.util.StringUtil;
+import seedu.address.logic.commands.FindCommand;
```
Comment (sumomomomomo) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:12]: Don't couple FindCommand to ParseUtil

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 9]
```diff
@@ -1,13 +1,15 @@
 package seedu.address.logic.parser;
 
 import static java.util.Objects.requireNonNull;
+import static seedu.address.logic.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
```
Comment (sumomomomomo) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:4]: Ditto, handle this in FindCommand

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 10]
```diff
@@ -58,7 +58,7 @@ public void parse_validArgsWithClassId_returnsFindCommand() {
     }
 
     @Test
```
Comment (sumomomomomo) [inline src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:60]: All of your testcases should follow the underscore naming scheme, not camelcase

Comment (hooyushing) [reply src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:60]: fixed

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java (hunk 5 above)
Comment (hooyushing) [reply src/test/java/seedu/address/logic/commands/FindCommandTest.java]: fixed

**Pull request merged**